libs/datamodel/core/src/transform/ast_to_dml/lift.rs

```diff
@@ -32,101 +30,120 @@ impl<'a> LiftAstToDml<'a> {
         let mut model = dml::Model::new(ast_model.name.name.clone(), None);
         model.documentation = ast_model.documentation.clone().map(|comment| comment.text);
 
-        let mut errors = Diagnostics::new();
+        let active_connector = self.db.active_connector();
+        let mut field_ids_for_sorting: HashMap<&str, ast::FieldId> = HashMap::with_capacity(ast_model.fields.len());
+
+        for (field_id, scalar_field_type) in self.db.iter_model_scalar_fields(model_id) {
+            let ast_field = &ast_model[field_id];
+            let arity = self.lift_field_arity(&ast_field.arity);
+            let mut attributes = Vec::with_capacity(ast_field.attributes.len());
+
+            let field_type =
+                lift_scalar_field_type(ast_field, scalar_field_type, &mut attributes, &self.db, diagnostics);
 
-        for ast_field in &ast_model.fields {
-            match self.lift_field(ast_field) {
-                Ok(field) => model.add_field(field),
-                Err(mut err) => errors.append(&mut err),
+            attributes.extend(ast_field.attributes.iter().cloned());
+
+            let mut field = dml::ScalarField::new(&ast_field.name.name, arity, field_type);
+            field.documentation = ast_field.documentation.clone().map(|comment| comment.text);
+            let mut field = dml::Field::ScalarField(field);
+
+            if let Err(mut errs) = self.attributes.field.validate_and_apply(&attributes, &mut field) {
+                diagnostics.append(&mut errs);
             }
+            field_ids_for_sorting.insert(&ast_field.name.name, field_id);
+            model.add_field(field);
         }
 
-        if let Err(mut err) = self.attributes.model.validate_and_apply(ast_model, &mut model) {
-            errors.append(&mut err);
+        for (field_id, relation_field) in self.db.iter_model_relation_fields(model_id) {
+            let ast_field = &ast_model[field_id];
+            let arity = self.lift_field_arity(&ast_field.arity);
+            let target_model = &self.db.ast()[relation_field.referenced_model].as_model().unwrap();
+            let relation_info = dml::RelationInfo::new(&target_model.name.name);
+
+            let mut field = dml::RelationField::new(&ast_field.name.name, arity, arity, relation_info);
+
+            field.supports_restrict_action(
+                active_connector.supports_referential_action(dml::ReferentialAction::Restrict),
+            );
+            field.emulates_referential_actions(active_connector.emulates_referential_actions());
+
+            field.documentation = ast_field.documentation.clone().map(|comment| comment.text);
+            let mut field = dml::Field::RelationField(field);
+
+            if let Err(mut errs) = self.attributes.field.validate_and_apply(ast_field, &mut field) {
+                diagnostics.append(&mut errs);
+            }
+
+            field_ids_for_sorting.insert(&ast_field.name.name, field_id);
+            model.add_field(field)
         }
```

Comment:
Seeking feedback on these two loops for scalar and relation fields. These could easily live in methods with lots of parameters. I don't know what is more readable. I would assume the loop body will tend to get _shorter_ with time.

Reply:
No, just keep them like this.